etna/articles/models.py

```diff
@@ -399,6 +399,62 @@ def get_datalayer_data(self, request: HttpRequest) -> Dict[str, Any]:
         )
         return data
 
+    @cached_property
+    def similar_items(
+        self,
+    ) -> Tuple[Union["ArticlePage", "FocusedArticlePage", "RecordArticlePage"], ...]:
+        """
+        Returns a maximum of three ArticlePages that are tagged with at least
+        one of the same ArticleTags. Items should be ordered by the number
+        of tags they have in common.
+        """
+        if not self.article_tag_names:
+            # Avoid unncecssary lookups
+            return ()
+
+        tag_ids = self.tagged_items.values_list("tag_id", flat=True)
+        if not tag_ids:
+            # Avoid unncecssary lookups
+            return ()
+
+        # Identify other live pages with tags in common
+        related_tags = (
+            Page.objects.live()
+            .public()
+            .not_page(self)
+            .exact_type(ArticlePage, FocusedArticlePage, RecordArticlePage)
+            .filter(tagged_items__tag_id__in=tag_ids)
```

Comment:
Do we know if there is any optimisation to be gained by changing this order? I'm just thinking that this is run against all pages on every page load (I think?)

If it does have an impact then maybe we should be applying the filter that omits most pages at the start so that subsequent filters have less and less to sift through?

Reply:
Hey AJ, sorry it took me a while to get back to you on this one - I _think_ what I've just pushed should make it a bit better, but tbh it's very hard to track the performance time with the Django Debug Toolbar, because it keeps changing the results. Running the same query 5 times gave me 5 very different numbers 😂